aquadoggo/src/validation.rs

```diff
@@ -0,0 +1,335 @@
+/// - ensures the claimed sequence number is equal to expected one  
+pub fn is_next_seq_num(latest_seq_num: Option<&SeqNum>, claimed_seq_num: &SeqNum) -> Result<()> {
+    let expected_seq_num = match latest_seq_num {
+        Some(seq_num) => {
+            let mut seq_num = seq_num.to_owned();
+            increment_seq_num(&mut seq_num)
+        }
+        None => Ok(SeqNum::default()),
+    }?;
+
+    ensure!(
+        expected_seq_num == *claimed_seq_num,
+        anyhow!(
+            "Entry's claimed seq num of {} does not match expected seq num of {} for given author and log",
+            claimed_seq_num.as_u64(),
+            expected_seq_num.as_u64()
+        )
+    );
+    Ok(())
+}
+
+/// Verify that an entry's claimed log id matches what we expect from the claimed document id.
+///
+/// This method handles both the case where the claimed log id already exists for this author
+/// and where it is a new log.
+///
+/// The following steps are taken:
+/// - Retrieve the stored log id for the claimed document id
+///   - If found, ensure it matches the claimed log ig
+///   - If not found retrieve the next available log id for this author and ensure that matches
+pub async fn verify_log_id<S: StorageProvider>(
+    store: &S,
+    author: &Author,
+    claimed_log_id: &LogId,
+    document_id: &DocumentId,
+) -> Result<()> {
+    // Check if there is a log_id registered for this document and public key already in the store.
+    match store.get(author, document_id).await? {
+        Some(expected_log_id) => {
+            // If there is, check it matches the log id encoded in the entry.
+            ensure!(
+                *claimed_log_id == expected_log_id,
+                anyhow!(
+                    "Entry's claimed log id of {} does not match expected log id of {} for given author and document",
+                    claimed_log_id.as_u64(),
+                    expected_log_id.as_u64()
+                )
+            );
+        }
+        None => {
+            // If there isn't, check that the next log id for this author matches the one encoded in
+            // the entry.
+            let expected_log_id = next_log_id(store, author).await?;
+
+            ensure!(
+                *claimed_log_id == expected_log_id,
+                anyhow!(
+                    "Entry's claimed log id of {} does not match expected next log id of {} for given author",
+                    claimed_log_id.as_u64(),
+                    expected_log_id.as_u64()
+                )
+            );
+        }
+    };
+    Ok(())
+}
+
+// Get the _expected_ skiplink for the passed entry.
+//
+// This method retrieves the expected skiplink given the author, log and seq num
+// of an entry. It _does not_ verify that it matches the claimed skiplink
+// encoded on the passed entry.
+//
+// If the expected skiplink could not be found in the database an error is returned.
+pub async fn get_expected_skiplink<S: StorageProvider>(
```

Comment:
Yeh, that is intentional. I wanted to keep each method simple, which means making certain assumptions, here I assume we already know that a skiplink is _expected_. Nice point though, I'll add a comment.

Reply:
Hmm, looking again, is what you say true though? I think `seq_num.skiplink_seq_num()` returns a lipmaa link whether it is required or not. So I don't think this method errors when it is not needed.

Reply:
Well, yeah, _if_ `SeqNum::skiplink_seq_num()` was working it would error unexpectedly :D Because the `match expected_skiplink_seq_num` would return `None` for both the case where no skiplink is needed and when it is needed but not found in the store.

Reply:
@cafca you happy with leaving this as is in this PR? Seems we have a bug elsewhere which needs sorting before this can be improved: https://github.com/p2panda/p2panda/issues/417

Reply:
(sorry, didn't see your comment when I wrote mine)

Reply:
> Well, yeah, _if_ `SeqNum::skiplink_seq_num()` was working it would error unexpectedly :D Because the `match expected_skiplink_seq_num` would return `None` for both the case where no skiplink is needed and when it is needed but not found in the store.

Is that the behaviour we want? To return `None` when a skiplink isn't required? We do accept skiplinks even when not required, right?

Reply:
I stumbled upon exactly this in the refactoring PR .. hmn!

> We do accept skiplinks even when not required, right?

Yes, the validation logic accepts that currently: https://github.com/p2panda/p2panda/blob/main/p2panda-rs/src/entry/entry.rs#L191 I've refactored it so it also errors right now, but doesn't need to be like that.

Reply:
I have a suggestion for this and will make a pr

Reply:
Thank you! 🙏🙏

Reply:
Please take a look and let me know what you think :) #218

Reply:
Nice improvement, thanks :pray: